src/squidpy/gr/_niche.py

```diff
@@ -0,0 +1,432 @@
+from __future__ import annotations
+
+from collections.abc import Iterator
+from typing import Any
+
+import anndata as ad
+import numpy as np
+import pandas as pd
+import scanpy as sc
+import scipy.sparse as sps
+from anndata import AnnData
+from scipy.sparse import hstack, issparse, spdiags
+from sklearn import metrics
+from sklearn.mixture import GaussianMixture
+from sklearn.preprocessing import normalize
+from spatialdata import SpatialData
+
+from squidpy._utils import NDArrayA
+
+__all__ = ["calculate_niche"]
+
+
+def calculate_niche(
+    adata: AnnData | SpatialData,
+    flavor: str = "neighborhood",
+    library_key: str | None = None,  # TODO: calculate niches on a per-slide basis
+    table_key: str | None = None,
+    mask: pd.core.series.Series = None,
+    groups: str | None = None,
+    n_neighbors: int = 15,
+    resolutions: int | list[float] | None = None,
+    subset_groups: list[str] | None = None,
+    min_niche_size: int | None = None,
+    scale: bool = True,
+    abs_nhood: bool = False,
+    adj_subsets: int | list[int] | None = None,
+    aggregation: str = "mean",
+    n_components: int = 3,
+    random_state: int = 42,
+    spatial_connectivities_key: str = "spatial_connectivities",
+) -> AnnData | pd.DataFrame:
+    """Calculate niches (spatial clusters) based on a user-defined method in 'flavor'.
+    The resulting niche labels with be stored in 'adata.obs'. If flavor = 'all' then all available methods
+    will be applied and additionally compared using cluster validation scores.
+    Parameters
+    ----------
+    %(adata)s
+    flavor
+        Method to use for niche calculation. Available options are:
+            - `{c.NEIGHBORHOOD.s!r}` - cluster the neighborhood profile.
+            - `{c.UTAG.s!r}` - use utag algorithm (matrix multiplication).
+            - `{c.CELLCHARTER.s!r}` - cluster adjacency matrix with Gaussian Mixture Model (GMM) using CellCharter's approach.
+            - `{c.SPOT.s!r}` - calculate niches using optimal transport. (coming soon)
+            - `{c.BANKSY.s!r}`- use Banksy algorithm. (coming soon)
+    %(library_key)s
```

Comment:
this won' twork without injecting docs, see the read the docs build to see how it renders

Reply:
but the function is not added in the docstrings so it is not rendering it at allk